smtp.go

```diff
@@ -2,209 +2,391 @@ package maddy
-	return nil, nil // TODO
-}
+		endp.listeners = append(endp.listeners, l)
 
-type keyRing struct {
-	pgppubkey.Source
-}
+		endp.listenersWg.Add(1)
+		addr := addr
+		go func() {
+			if err := endp.serv.Serve(l); err != nil && !strings.HasSuffix(err.Error(), "use of closed network connection") {
+				endp.Log.Printf("failed to serve %s: %s", addr, err)
+			}
+			endp.listenersWg.Done()
+		}()
+	}
 
-func (kr *keyRing) Unlock(username, password string) (openpgp.EntityList, error) {
-	return pgplocal.Unlock(username, password)
-}
+	if lmtpUsed {
+		endp.serv.LMTP = true
+	}
 
-func newSMTPPGP(d caddyfile.Dispenser, be smtp.Backend) (smtp.Backend, error) {
-	kr := &keyRing{pgppubkeylocal.New()}
-	pgpbe := smtppgp.New(be, kr)
-	return pgpbe, nil
+	return nil
 }
 
-func newVerifier(be smtp.Backend) smtp.Backend {
-	identity := "localhost"
-
-	return &backendutil.TransformBackend{
-		Backend: be,
-		Transform: func(from string, to []string, r io.Reader) (string, []string, io.Reader, error) {
-			var b bytes.Buffer
-			if _, err := io.Copy(&b, r); err != nil {
-				return "", nil, nil, err
-			}
-
-			results, err := verifyDKIM(bytes.NewReader(b.Bytes()))
+func (endp *SMTPEndpoint) setDefaultPipeline(localDeliveryName, remoteDeliveryName string, localOpts, remoteOpts map[string]string) error {
+	var err error
+	var localDelivery module.DeliveryTarget
+	if localDeliveryName == "" {
+		localDelivery, err = deliveryTarget("default_local_delivery")
+		if err != nil {
+			localDelivery, err = deliveryTarget("default")
 			if err != nil {
-				return "", nil, nil, err
+				return err
 			}
+		}
+		localOpts = map[string]string{"local_only": ""}
+	} else {
+		localDelivery, err = deliveryTarget(localDeliveryName)
+		if err != nil {
+			return err
+		}
+	}
 
-			// TODO: strip existing Authentication-Results header fields with our identity
+	if endp.submission {
+		if remoteDeliveryName == "" {
+			remoteDeliveryName = "default_remote_delivery"
+			remoteOpts = map[string]string{"remote_only": ""}
+		}
+		remoteDelivery, err := deliveryTarget(remoteDeliveryName)
+		if err != nil {
+			return err
+		}
 
-			// TODO: don't print Authentication-Results on a single line
-			authRes := "Authentication-Results: " + msgauth.Format(identity, results) + "\r\n"
-			r = io.MultiReader(strings.NewReader(authRes), &b)
-			return from, to, r, nil
-		},
+		endp.pipeline = append(endp.pipeline,
+			// require_auth and submission_check are always prepended to pipeline
+			//TODO: DKIM sign
+			deliveryStep{t: localDelivery, opts: localOpts},
+			deliveryStep{t: remoteDelivery, opts: remoteOpts},
+		)
+	} else {
+		endp.pipeline = append(endp.pipeline,
+			//TODO: DKIM verify
+			deliveryStep{t: localDelivery, opts: localOpts},
+		)
 	}
+
+	return nil
+}
+
+func (endp *SMTPEndpoint) Login(state *smtp.ConnectionState, username, password string) (smtp.User, error) {
+	if !endp.Auth.CheckPlain(username, password) {
```

Comment:
We could probably make this fail if it's a submission endpoint

Reply:
This is handled by prepending `require_auth` step to the pipeline in "submission mode".